shell/common/input_events_unittests.cc

```diff
@@ -0,0 +1,227 @@
+
+  ASSERT_GE(base_latency, 0);
+
+  // Check that delivery_time satisfies our assumptions.
+  int continuous_frame_count = 0;
+  for (int i = 0; i < num_events; i += 1) {
+    // j is the frame index of event i if there were no latency.
+    int j = static_cast<int>((delivery_time(i) - base_latency) / frame_time);
+    if (j == continuous_frame_count) {
+      continuous_frame_count += 1;
+    }
+    double random_latency = delivery_time(i) - j * frame_time - base_latency;
+    ASSERT_GE(random_latency, 0);
+    ASSERT_LT(random_latency, frame_time);
+
+    // If there were no latency, there should be at least one event per frame.
+    // Hence j should never skip any integer less than continuous_frame_count.
+    ASSERT_LT(j, continuous_frame_count);
+  }
+
+  // i is the input event's index.
+  // j is the frame's index.
+  for (int i = 0, j = 0; i < num_events; j += 1) {
+    double t = j * frame_time;
+    while (i < num_events && delivery_time(i) <= t) {
+      ShellTest::DispatchFakePointerData(shell.get());
+      i += 1;
+    }
+    ShellTest::PumpOneFrame(shell.get());
+  }
+
+  shell.reset();
+}
+
+TEST_F(ShellTest, MissAtMostOneFrameForIrregularInputEvents) {
+  // We don't use `constexpr int frame_time` here because MSVC doesn't handle
+  // it well with lambda capture.
+  int frame_time = 10;
+  int base_latency = 0.5 * frame_time;
+  Generator extreme = [frame_time, base_latency](int i) {
+    return static_cast<int>(i * frame_time + base_latency +
+                            (i % 2 == 0 ? 0.1 * frame_time : 0.9 * frame_time));
+  };
+  constexpr int n = 40;
+  std::vector<int> events_consumed_at_frame;
+  TestSimulatedInputEvents(this, n, base_latency, extreme, frame_time,
+                           events_consumed_at_frame);
+  int frame_drawn = events_consumed_at_frame.size();
+  ASSERT_GE(frame_drawn, n - 1);
+}
+
+TEST_F(ShellTest, DelayAtMostOneEventForFasterThanVSyncInputEvents) {
+  // We don't use `constexpr int frame_time` here because MSVC doesn't handle
+  // it well with lambda capture.
+  int frame_time = 10;
+  int base_latency = 0.2 * frame_time;
+  Generator double_sampling = [frame_time, base_latency](int i) {
+    return static_cast<int>(i * 0.5 * frame_time + base_latency);
+  };
+  constexpr int n = 40;
+  std::vector<int> events_consumed_at_frame;
+  TestSimulatedInputEvents(this, n, base_latency, double_sampling, frame_time,
+                           events_consumed_at_frame);
+
+  // Draw one extra frame due to delaying a pending packet for the next frame.
+  int frame_drawn = events_consumed_at_frame.size();
+  ASSERT_EQ(frame_drawn, n / 2 + 1);
+
+  for (int i = 0; i < n / 2; i += 1) {
+    ASSERT_GE(events_consumed_at_frame[i], 2 * i - 1);
+  }
+}
+
+TEST_F(ShellTest, HandlesActualIphoneXsInputEvents) {
+  // Actual delivery times measured on iPhone Xs, in the unit of frame_time
+  // (16.67ms for 60Hz).
+  constexpr int n = 47;
+  constexpr double iphone_xs_times[n] = {0.15,
```

Comment:
You can get rid of `n = 47` and instead write `sizeof(iphonx_xs_times) / sizeof(double)` or `arraysize`. That way, we can add and remove items from this fixture without potentially forgetting to update the size. These are known at compile time anyway.

Reply:
Done.